### DISCUSSION TITLE
Allow shared reusable embeddable "blocks" across documents

### DISCUSSION THREAD

**Discussion by designatedsuccessor** [outline/outline#7164] [Ideas]
We currently have a contacts page that contains all contact info for users.  When we want to refer to someone's contact info, we link to the contacts page.

Users really want to put contact info directly in the page, instead of always linking to the contacts page.  But the obvious disadvantage of doing that is that once someone's contact info updates, all instances of their contact info must be updated across the entire Outline instance.

We'd like to have a page within the Outline settings area where we can create reusable shared blocks of data.  Example block:

```
Name:
Vice President's Contact Info

Content:
John Smith
555-555-5555
johnsmith@example.com

Allowed Within These Collections:
Collection A
Collection D
Collection V

Permissions:
User Group A: Can Edit (And Use in Documents)
User Group C: Can Use in Documents Only
User F: Can Edit (And Use in Documents)
```

Imagine it was embedded as so:

> Thank you for your interest.  The Vice President's contact information is as follows:
> 
> [_Vice President's Contact Info_]
> 
> If you have any questions, please contact the Vice President as soon as possible.

To the end user, it would look like this:

> Thank you for your interest.  The Vice President's contact information is as follows:
> 
> John Smith
> 555-555-5555
> johnsmith@example.com
> 
> If you have any questions, please contact the Vice President as soon as possible.

**Comment by designatedsuccessor** [comment]
@tommoor I know you're tired of me by now, but here are some mock-ups of what I think this blocks/modular components idea could look like:

Example of collections and associated modular components.  So it would have (a) the name of each collection appear within the "Collections" section just like it does today and (b) the name of each collection within the new "Modular Components" section *if* the collection has modular components enabled; collections in which modular components were disabled would not appear within the "Modular Components" section.
![image](https://github.com/user-attachments/assets/8a1e331f-a50c-438a-ace3-99a0fec9e4a8)

Example of enabling/disabling modular components for a collection:
![image](https://github.com/user-attachments/assets/3b54e29c-68cf-468e-9321-aefdf00581e2)

Example of the contents of a modular component:
![image](https://github.com/user-attachments/assets/267aaa66-9ccb-4aac-9355-00137dfb960c)

Example of process of adding a modular component:
![image](https://github.com/user-attachments/assets/656e5983-d7fb-4e21-a8e8-0cdc86f4e4ce)

Example of after a modular component has been added to a document, while in "Edit" mode (with a "disconnect" button on the right side to essentially delete it from the current document without deleting the modular component itself).  Of course, when users see it in non-edit mode, it will show the actual contents of the modular component.
![image](https://github.com/user-attachments/assets/2af1bafd-7cea-4dce-8ee1-c5841e2ad9dd)

Deletion expected behavior: if you had a modular component added to, say, 3 documents and then you completely deleted the modular component, then I think it would be expected that the modular component would no longer appear in any of the 3 documents.  To help with processing speeds/times, maybe at the exact moment the modular component was deleted, the database reference would continue existing for the time being except it would be marked for deletion, so that when a document was opened that contained the modular component, the document would know not to show anything to the user where the modular component is still located.  Then, a once-a-day batch process could go through and clean up (a) the database reference and (b) all the documents that contain the "behind the scenes" version of the modular component.

**Comment by EduNetEurope** [comment]
We would love to see reusable content as well.
There is another request which refers to resuable content as snippets: https://github.com/outline/outline/discussions/2030

**Comment by AndrewBucklin** [comment]
@designatedsuccessor - I really like your mock ups!

This would be very helpful for our team. We manage multiple clients and we tend to reuse certain steps / processes for each client. Sometimes these steps charge or need to be updated and we constantly find ourselves following various revisions of those steps since we tend to forget to update the steps manually in all located throughout the Wiki. One workaround we have tried to implement is storing those common steps in a Google Doc and linking to that (or embedding it), but it's clunky and goes against the idea of having all documentation within a single system.

**Comment by janstadt** [comment]
Just curious if this feature is being developed? I see two issues submitted and they both link back to one another but there havent been any updates for quite a while.

**Reply by AndrewBucklin** [reply]
I would be willing to toss some cash to one of the devs to work on this using a site such as [bountyhub.dev](https://www.bountyhub.dev/en)

**Reply by designatedsuccessor** [reply]
@AndrewBucklin we likely would, too.  It's important for us.

**Comment by designatedsuccessor** [comment]
@tommoor any chance that this can still be a developed feature?

I was thinking that, for performance, could a block of data be hard-coded into referenced pages and updated only when the user clicks a button?

Example: instead of a page containing this:

```
This is the VP's contact data:

<Reference to Contact Block>
```

It would contain:

```
This is the VP's contact data:

<Begin Contact Block>
Phone: 555-555-5555
Email: john@contoso.com
<End Contact Block>
```

That way, visiting a page with, say, 10 data blocks would not result in 10 lookups.  But the page would still contain a reference to the data block so that when it got updated, the updating process would know where it was located.

And instead of each data block updating the references, the user would need to click an Update button each time they wanted the associated references to be updated.  So data-block updates should occur less often.

And it could go even one further: a batch process that runs once per day to update data-block references.  That way, if a user updated a block 200 times in a day, the block's associated references would be updated only once that day.

**Reply by tommoor** [reply]
I'm not totally against it, I do believe this feature will exist in some form eventually. I have to work on some non-editor stuff for a while though

**Reply by designatedsuccessor** [reply]
Is there anyone you trust with it, who would work it for some level of cash?

**Reply by AndrewBucklin** [reply]
We are still here and still really wishing this feature would be developed. And still open to throwing some $$ at the request.

I think I mentioned it before, but we really need a way to globalize shared processes. Right now, updating a single step requires manual changes across dozens of Wiki pages, leading to inevitable human error and 'version drift.' Our current workaround is embedded Google Docs, but that's a clunky fix that pulls us out of our primary system. A native solution for reusable content blocks would be a game-changer for our team.  🙌

**Reply by dorianvee** [reply]
Same here ... will also throw in some $$. 

FYI .. check https://www.notion.com/help/synced-blocks

This is exactly what we need in outline
